### PR TITLE
Upgrade test case 41

### PR DESCRIPTION
**Pull request by dylant-da**

### CODE REVIEW

File: sdk/daml-script/test/daml/upgrades/DataChanges.daml [hunk 1]
```diff
@@ -286,3 +317,336 @@ templateEnumDowngradeFromOld =
...
+
+templateOptionalFieldOnChoiceSetWhenDowngrading : Test
+templateOptionalFieldOnChoiceSetWhenDowngrading = test $ do
+  a <- allocatePartyOn "alice" participant0
+  cid <- a `submit` createExactCmd V1.OptionalFieldOnChoiceAdded with party = a
+
+  let cidV2 = coerceContractId @V1.OptionalFieldOnChoiceAdded @V2.OptionalFieldOnChoiceAdded cid
+
+  -- Ensure we can only call the V1 choice
+  (actAs a <> packagePreference [dataChangesV1]) `trySubmit` exerciseCmd cidV2 V2.OptionalFieldOnChoiceAddedCall { existingField = "V2", newField = Some "V2" }
```
Comment (paulbrauner-da) [inline sdk/daml-script/test/daml/upgrades/DataChanges.daml:463]: I think this line is redundant?

File: sdk/daml-script/test/daml/upgrades/DataChanges.daml [hunk 2]
```diff
@@ -286,3 +317,336 @@ templateEnumDowngradeFromOld =
...
+    b <- allocatePartyOn "bob" participant0
+    res <- f a b
+    case res of
+      Left (DevError Upgrade msg) | "An optional contract field with a value of Some may not be dropped during downgrading" `isInfixOf` msg -> pure ()
+      _ -> assertFail $ "Expected specific failure but got " <> show res
+
+tofswdViaInterfaceInCommandGlobal : (Text, Test)
+tofswdViaInterfaceInCommandGlobal = templateOptionalFieldsSetWhenDowngradingTest "dynamic, global,    in command" $ \a b -> do
+  cid <- a `submit` createCmd V2.OptionalFieldsAddedUsingIface with party = a, newField = Some "V1"
+  let cidIface = coerceContractId @V2.OptionalFieldsAddedUsingIface @Iface.I cid
```
Comment (paulbrauner-da) [inline sdk/daml-script/test/daml/upgrades/DataChanges.daml:558]: Nit: for the special case of coercing to an interface ID, there's a toInterfaceContractId function (which, as far as I understand, does the same thing but statically checks that this is safe, i.e. that the template has an instance for that interface).

**Review by paulbrauner-da — APPROVED** [review]
Congrats on completing every single column :)

File: sdk/daml-script/test/daml/upgrades/DataChanges.daml (hunk 1 above)
Comment (dylant-da) [reply sdk/daml-script/test/daml/upgrades/DataChanges.daml]: Ah yes, I noticed this while porting to 2.x, deleted there, deleted here now :)

File: sdk/daml-script/test/daml/upgrades/DataChanges.daml [hunk 3]
```diff
@@ -286,3 +317,336 @@ templateEnumDowngradeFromOld =
...
+              (Some "V1") -- @V  2
+            )
+          let cidIface = coerceContractId @OptionalFieldsAddedUsingIface @I cid
+          exercise cidIface (NoOp party)
+-}
+
+dataChangesV1 : PackageId
+dataChangesV1 = getPackageId "data-changes-1.0.0"
+
+templateOptionalFieldsSetWhenDowngradingTest : Show b => Text -> (Party -> Party -> Script (Either SubmitError b)) -> (Text, Test)
```
[the remark this answers is not in the document]
Comment (dylant-da) [reply sdk/daml-script/test/daml/upgrades/DataChanges.daml:509]: Moved, along with a structural refactor which combines some things into one module where possible and renames generically-named things like Iface.I to OptionalFieldsAddedIface

File: sdk/daml-script/test/daml/upgrades/DataChanges.daml (hunk 2 above)
Comment (dylant-da) [reply sdk/daml-script/test/daml/upgrades/DataChanges.daml]: fixed

**Pull request merged**